print(plus.garbage.count())
print(...)
plus.io.stderr:print("helllo"):ln()
plus.io.stdout:println("eee"):println("eee"):print("eeeee"):print("aaaa")
plus.io.stderr:print("eeeeea")
print(`garbage: {pls.gc.count()}kb`)
print(typeof(plus.filesystem.path("")))
-- for child in plus.filesystem.current_path():child("-cmake"):iterate_descendants() do
--     print(child:string())
-- end
plus.filesystem.path("ee")
print(`garbage: {plus.garbage.count()}kb`)
plus.garbage.collect()
print(`garbage: {plus.garbage.count()}kb`)

-- plus.io.file_writer("abc", true):println("im chanining"):close_after(function(self)
--     print("closing")
-- end)
local fw = pls.io.file_writer("abc", true)
type writer = plus_io_writer
local writer: writer = fw:println("eeeee"):ln():ln()
fw:println("eeee"):print("eeee"):ln()
local cwd = plus.filesystem.current_path()
print(cwd)
local m = require("@test")
print(m.name)
print(plus.garbage.count())
